### PR TITLE
feat(npm): extract rules from npmrc [WIP]

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds function to extract `hostRules` and `packageRules` from `npmrc` content.

## Context:

Required as part of #10110 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Draft PR, not yet used in the code. @viceice any patterns I've missed in `npmrc` parsing?

File: lib/config/npmrc.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      const hostPart = key.replace(/:_authToken$/, '');
+      res.hostRules ||= [];
+      res.hostRules.push({
+        hostType,
+        matchHost: getMatchHost(hostPart),
+        token: val,
+      });
+      // packageRules
+    } else if (key === 'registry') {
+      const [, host] = val.split('//');
```
Comment (viceice) [inline lib/config/npmrc.ts:46]: Why not directly use val? It must be a url.

File: lib/config/npmrc.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        matchHost: getMatchHost(hostPart),
+        authType: 'Basic',
+        token: val,
+      });
+    } else if (key.endsWith(':_authToken')) {
+      const hostPart = key.replace(/:_authToken$/, '');
+      res.hostRules ||= [];
+      res.hostRules.push({
+        hostType,
+        matchHost: getMatchHost(hostPart),
```
Comment (viceice) [inline lib/config/npmrc.ts:41]: Hostpart always starts with `//` for npmrc. `http` or `https` is defined by registry url setting.

**Review by viceice — COMMENTED** [review]
We shouldn't support multiple overlapping auth settings per path.

**Comment by rarkins** [comment]
> We shouldn't support multiple overlapping auth settings per path.

Won't the later/second one take precedence anyway?

**Comment by viceice** [comment]
> > We shouldn't support multiple overlapping auth settings per path.
> 
> Won't the later/second one take precedence anyway?

Yes, i think so, that's why i would like to explicit mention it in readme. So poeple won't expect this to work.

File: lib/config/npmrc.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          matchDatasources: [hostType],
+          registryUrls: [val],
+        });
+      }
+    } else if (key.endsWith(':registry')) {
+      if (validateUrl(val, false)) {
+        const [scope] = key.split(':');
+        res.packageRules.push({
+          matchDatasources: [hostType],
+          matchPackagePrefixes: [scope],
```
Comment (viceice) [inline lib/config/npmrc.ts:49]:
```suggestion
          matchPackagePrefixes: [ensureTrailingSlash(scope)],
```
So we don't mix `@org` and `@org-stuff` 🙃

**Pull request closed without merging**